### PR TITLE
CXXCBC-885: hand-rolled test harness for the CNG test tree

### PR DESCRIPTION
**Pull request by avsej**
CNG work needs a test harness that is not Catch2, the couchbase-cxx-client
default. Add a small framework under `test/cng/framework/`, ported to the C++17
the library targets (the source design is C++23): a `tests()` entry point
returning a `test_suite` of `test_case{name, func, timeout, env}`; a per-case
timeout runner that reports pass/skip/fail without aborting the process;
environment gating (agnostic/any_server/mock_only/cluster_only) driven by
`TEST_CONNECTION_STRING`; and `assert_*` helpers over a `source_location` shim
built on compiler builtins.

Gated by `COUCHBASE_CXX_CLIENT_BUILD_CNG_TESTS` and registered with ctest via
`add_cng_test()`; the harness does not link the client library and leaves the
Catch2 suite untouched. A self-test exercises the pass, fail, skip, timeout,
gating, and filter paths.

---
Part of the CNG (`couchbase2://`) transport series — stacked PR **01/29**, the base of the stack. Merge bottom-up.

### CODE REVIEW

File: test/cng/framework/test_runner.cxx [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+    } catch (...) {
+      return { case_result::status::failed, duration, "threw an unknown exception" };
+    }
+  }
+
+  worker.detach();
+  const auto duration = std::chrono::steady_clock::now() - start;
+  return { case_result::status::failed,
+           duration,
+           fmt::format("timed out after {}", human(timeout)) };
```
[the remark this answers is not in the document]
Comment (avsej) [reply test/cng/framework/test_runner.cxx]: The detach is deliberate: a C++ thread running arbitrary test code cannot be safely cancelled, so on timeout we detach rather than block teardown. The worker writes only to the packaged_task's heap-owned shared state (never the test's stack frame) and the timed-out result is simply never read (see the comment on run_case). Keeping as-is.

**Comment by thejcfactor** [comment]
🤖 This review was produced with Claude Code and posted on my behalf. Two findings were verified with isolated single-file compiles (the `source_location` shim behaviour and the `assert_eq` suggestion below); everything else is a static read of the diff — I did not build the CNG tree or run the suite. Please treat the reasoning as a starting point for discussion, not ground truth.

---

Reviewed at `8812b967`, scoped to this PR's commit only. The harness reads well — the C++17 backport is careful, the self-test covers the paths that matter, and the isolation (no client library linked, Catch2 suite untouched, gated behind `COUCHBASE_CXX_CLIENT_BUILD_CNG_TESTS`) is exactly right for landing at the base of a long stack. Comments are unusually good at explaining *why*, which made this easy to follow.

Two items I think are worth acting on, then some smaller ones.

## 1. `assert_eq` discards both values

`test_framework.hxx:227-235` — the failure message is `file:line: message`, where `message` defaults to the literal `"expected equal"`. The compared values never make it out.

This is the most-used assertion in the harness, and it's the one place where dropping Catch2 costs something concrete: expression decomposition is Catch2's headline feature. The self-test shows the cost on itself — when

```cpp
assert_eq(r.passed, std::size_t{ 1 }, "one pass counted");   // framework_selftest.cxx:86
```

fails in CI, you learn that `r.passed != 1` but not whether it was 0, 2, or 7. Diagnosing means editing the test, adding prints, and re-running.

`fmt::is_formattable` is available in the bundled fmt (10.2.1), so this can be gated at compile time without constraining what types `assert_eq` accepts:

```cpp
if (!(actual == expected)) {
  if constexpr (fmt::is_formattable<A>::value && fmt::is_formattable<B>::value) {
    throw test_assertion_failure(fmt::format("{}:{}: {} (actual: {}, expected: {})",
                                             loc.file_name(), loc.line(), message,
                                             actual, expected));
  } else {
    throw test_assertion_failure(fmt::format("{}:{}: {}",
                                             loc.file_name(), loc.line(), message));
  }
}
```

I compiled this against spdlog's bundled fmt to check it holds up: `is_formattable<int>` is 1, a plain non-formattable struct is 0, formattable types report `(actual: 4, expected: 5)`, and non-formattable types still compile through the fallback.

## 2. Stop-at-first-failure hides every later failure in the same binary

`test_runner.cxx:120` sets `stop = true` on the first failure and abandons the remainder (noted as matching the source harness). What sharpens this is `CMakeLists.txt:36`: `add_cng_test` registers one ctest test **per executable**, not per case, and each executable runs its whole suite. So the first failing case suppresses every later case in that binary, and `slow_test_cases` never run at all (`:127`).

As the tree grows — a KV test file will accumulate a lot of cases over the next several PRs — a change that breaks cases 3 and 25 reports only case 3. Case 25 stays invisible until 3 is fixed and CI is re-run, so N regressions cost N round-trips.

Would you consider run-all as the default with stop-at-first behind a flag or env var? Happy to be told the fail-fast behaviour is deliberate for gateway tests, where cascading failures after a connection dies are just noise — that's a reasonable argument, and if so it might be worth saying in the comment.

## 3. A suite that executes nothing exits 0

`test_runner.cxx:134-144` — `exit_code` returns 77 when `passed == 0 && skipped > 0`, but `0 passed, 0 skipped, 0 failed` falls through to `0`. The filter is an exact-match `std::set<std::string>` (`:98`) with no diagnostic for names that match nothing.

So `./cng_framework_selftest_test typo_name` runs nothing and exits 0. More durably: if a future test file's `tests()` ends up empty, or every case is filtered out, ctest reports **PASS** for a binary that tested nothing — and stays green indefinitely. Treating "zero cases executed" as a failure, and erroring on filter names that matched nothing, would close both.

## 4. Timeout detaches the worker, and the process can exit under it

`test_runner.cxx:76`. The comment at `:50-52` is right that the `packaged_task` co-owns the future's shared state, so the detached thread never writes to `run_case`'s dead frame. The part not covered is process teardown: a timeout also sets `stop = true`, so `run()` returns straight away, `main` prints its summary and returns, and static destructors begin — while the detached body is still running.

Nothing in the tree can hit this harmfully today, since the only timing-out case is the self-test's *inner* `run()` and the outer suite outlives its 200ms sleeper. It becomes real once a case does I/O: a `cluster_only` KV test with `timeout::network` blocking on an unresponsive gateway would leave a thread holding a live gRPC channel, potentially touching spdlog's registry or `std::cout` as they're destroyed. The symptom would be a nondeterministic abort *at exit*, which is the outcome the timeout runner is designed to avoid — and sanitizer builds would also flag the leaked thread.

**Not blocking this PR** — it's latent until the I/O tests land, and it may well be cleaner to fix alongside them. Flagging it now mainly so it doesn't get rediscovered later as a mystery CI flake with no obvious link back to the harness. If you'd rather defer, could it be pinned to whichever ticket adds the first tests that actually reach a gateway (CXXCBC-891?) rather than left to a review comment?

For whenever it is addressed: since a timeout is currently recorded as a generic failure, `main` can't tell that a thread was detached. Threading a `timed_out` flag from `run_case` through `run_result` would let `main` do:

```cpp
if (result.timed_out > 0) {
  std::cout.flush();
  std::_Exit(exit_code(result));   // a detached test body may still be running
}
```

Doing this conditionally rather than always calling `_Exit` keeps normal exits intact — an unconditional `_Exit` would also suppress LSAN's atexit reporting, which you probably want in a test harness.

## 5. The shim's MSVC comment doesn't match its logic

`test_framework.hxx:45-58` — the comment says the builtins are exposed by "GCC/Clang (and MSVC 16.6+)", but the condition only enables them when `__has_builtin` is defined, or `__clang__` is. MSVC didn't provide `__has_builtin` until VS 2022 17.1, so on MSVC 16.6–17.0 — which per the comment does support `__builtin_FILE`/`LINE`/`FUNCTION` — the shim silently takes the `"unknown"` / line `0` branch.

It compiles fine, so the degradation is invisible; assertion failures on that toolchain would report `unknown:0`. Either add `_MSC_VER >= 1926` to the condition or narrow the comment to say the builtins are used only where `__has_builtin` exists. (Static observation — I have no MSVC to hand.)

## 6. `should_run`: "no connection string" vs. "a mock is available"

`test_framework.hxx:116-140` treats `real_cluster == false` as mock mode, but the flag only encodes that `TEST_CONNECTION_STRING` is unset. Since there's no mock infrastructure yet, `any_server` and `mock_only` cases would *execute* with nothing to talk to — likely failing on connection errors, or hanging to their timeout.

More a question about where this is heading than a defect: which PR brings in the mock, and is a third state worth having (no server at all → skip both), or is the intent that `mock_only` cases simply won't exist until the mock does?

## Nits

- `test_framework.hxx:110-111` — `timeout::integration` and `timeout::slow` are both `30'000ms`. Fine if intentional, but a reader may assume they differ, and changing one later would silently shift the meaning of cases that chose the other.
- `CMakeLists.txt:26-27` — the target name comes from `NAME_WE`, so `kv/options.cxx` and `query/options.cxx` would both claim `cng_options_test` and collide. Deriving from the full relpath would avoid it as the tree fills out.
- `test_runner.cxx:44-48` — `human()` truncates to whole milliseconds, so `timeout::instant`-class cases all report `0ms`, which makes a case that got 10× slower invisible. `duration<double, std::milli>` with one decimal would keep it cheap.
- `test_framework.hxx:240-257` — `assert_throws` converts any non-`Exc` throw into an assertion failure, so a `skip()` (or a nested assertion failure) inside the callable would be reported as "a different exception type was thrown" rather than propagating. Almost certainly fine in practice; noting it in case skip-inside-a-lambda is expected to work.

## Things I checked that look correct

- The `source_location` shim does what it claims. I compiled it standalone under `-std=c++17` to confirm the builtins-as-default-arguments trick captures the *caller's* location: three call sites reported three distinct line numbers with the correct enclosing function names, where a definition-site capture would have reported the same line three times.
- `safe_getenv` handles the MSVC `getenv` deprecation via `_dupenv_s` and normalizes unset *and* empty to `nullopt`, matching the existing wrappers.
- `SKIP_RETURN_CODE 77` (`CMakeLists.txt:38`) lines up with the harness's 77 (`test_runner.cxx:141`).
- The self-test asserting on `run()`'s *return value*, rather than letting inner failures propagate, is a nice touch — it exercises the fail and timeout paths while the binary still exits 0. All eight `should_run` combinations are covered.

**Comment by avsej** [comment]
Thanks — this is a genuinely useful review, and the two items you called out as worth acting on were both right. All six findings and all four nits are addressed; details and the one place I pushed back below.

One correction to the premise of #1: the bundled fmt here is **11.0.2**, not 10.2.1 (`FMT_VERSION 110002` in spdlog's vendored copy). Immaterial to your suggestion — I re-checked against the actual header and `fmt::is_formattable` is still there (`base.h:1995`), with `is_formattable<int>::value == 1`, a plain struct `0`, and the fallback compiling — but worth noting in case the version informed anything else.

## 1. `assert_eq` discards both values — fixed

Implemented as you suggested. Your own example now reports:

```
framework_selftest.cxx:86: one pass counted (actual: 7, expected: 1)
```

Non-formattable types take the fallback branch unchanged.

## 2. Stop-at-first-failure — fixed, run-all is now the default

You're right, and the `add_cng_test`-registers-one-test-per-executable observation is what makes it matter. I didn't put fail-fast behind a flag: nothing wants it. Each case is independent (its own connection, its own in-process server where applicable), so the cascading-noise argument doesn't really apply here, and ctest already isolates at the binary level. Deleting the `stop` flag was a smaller diff than making it configurable — happy to add the switch if a real need turns up. `slow_test_cases` now run unconditionally too.

## 3. A suite that executes nothing exits 0 — fixed, both halves

`exit_code` now returns 1 when nothing ran, nothing skipped and nothing failed, and an unmatched filter name is itself a failure. Your example:

```
$ ./cng_framework_selftest_test typo_name
No test case named "typo_name" in suite "framework_selftest"
Suite "framework_selftest": 0 passed, 0 skipped, 1 FAILED
$ echo $?
1
```

## 4. Timeout detaches the worker, and the process can exit under it — fixed now, not deferred

Your teardown analysis is correct, and I'd rather not leave it to be rediscovered as a CI flake. Done as you described: `run_result::timed_out` threaded from `run_case`, and `main` does a conditional `std::_Exit` — conditional exactly so LSAN's atexit report survives normal runs. It's harness-internal, so it belongs here rather than riding along with CXXCBC-891.

## 5. The shim's MSVC comment doesn't match its logic — comment fixed, `#if` left alone

Agreed the comment was wrong. I narrowed the comment rather than adding `_MSC_VER >= 1926`: CI has no MSVC 16.6–17.0 leg (the Windows legs are VS 2022 17.14, which has `__has_builtin`, so they already take the builtins path), and an untested `#if` for a toolchain nothing builds seemed worse than a documented limitation. The comment now states the gap explicitly.

## 6. `should_run`: "no connection string" vs. "a mock is available" — documentation was misleading; no mock is coming

Not a defect, but the comment was actively misleading and is rewritten. To answer directly: **no PR brings a mock.** Tests needing a server stand up an in-process gRPC server and are therefore `agnostic`. Across the whole stack only `agnostic` (3 uses) and `cluster_only` (13) appear — `any_server` and `mock_only` are unused, carried over from the harness this was ported from.

And you're right that `any_server` would execute with nothing to talk to. That's now called out in the comment, with the third state you suggest as the precondition for adopting it. I left the two values in place rather than deleting them, since they're part of the ported design — but if they're still unused when the stack lands, deleting them is the honest cleanup.

## Nits — all four addressed

- **`integration`/`slow` both 30s** — kept, with a comment that the equality is coincidental and either may be retuned independently.
- **`NAME_WE` collision** — added a configure-time `if(TARGET ${target})` → `FATAL_ERROR`. Deriving names from the full relpath would rename every existing target and churn the ctest names across all 29 commits; a loud failure the moment two stems collide buys the same safety for three lines.
- **`human()` truncation** — now `duration<double, std::milli>` at one decimal, so fast cases read `0.1ms` instead of `0ms`.
- **`assert_throws` swallowing `skip()`** — `test_skip_exception` and `test_assertion_failure` now rethrow instead of being reported as "a different exception type".

## Regression coverage

Five new self-test cases, so none of the above can silently regress: a failure not suppressing later cases, slow cases surviving an earlier failure, unmatched filter name failing, an empty suite not passing, and a timeout being flagged distinctly from an ordinary failure. Suite is 12/12; `ctest -L cng` is 34/34.

Also confirming your note on `safe_getenv` — that landed in this commit for exactly the reason you describe: MSVC's `/W4 /WX` turns `std::getenv`'s C4996 into C2220, which broke all three Windows legs. They're green now.

**Review by thejcfactor — APPROVED** [review]
